### PR TITLE
Added all timers for all variants as described by CubeMX

### PR DESCRIPTION
**Pull request by cs2dsb**
I've added all the timers according to the CubeMX database with the correct device & density flags.

The only one's I've had to comment out are 9, 10, 12, 13 and 14 on some devices because fields for these are missing from the underlying pac. The code is there to enable them when stm32-rs fixes up those SVDs.

This might be a minor breaking change as previously some of the devices had timers enabled that aren't actually available across the whole range - the fix is to add the appropriate density feature. I've added some info about this to the README so hopefully people will find this without a problem.

There's also some minor formatting stuff mostly from rustfmt in case you are wondering about the whitespace changes in pwm_input.

I should also draw your attention to the fact I added #![allow(unused_unsafe)] (with a comment) to lib.rs to hide the extremely numerous warnings that were coming up because some pac have safe fields incorrectly identified as unsafe. I think adding #![allow(unused_unsafe)] to the crate is fine and preffererable to it being harder to read the compiler output but I'm happy to be persuaded otherwise if you disagree.

### CODE REVIEW

File: src/timer.rs [hunk 1]
```diff
@@ -269,16 +292,75 @@ macro_rules! hal {
...
     TIM4: (tim4, pclk1_tim, dbg_tim4_stop),
 }
+
+#[cfg(feature = "HIGH")]
+hal! {
+    TIM5: (tim5, pclk2_tim, dbg_tim5_stop),
+}
+
+/*
+    dbg_tim(12-13)_stop fields missing from 103 xl in stm32-rs
```
Comment (geomatsi) [inline src/timer.rs]: Maybe leave FIXME or TODO mark here and in the next commented section to remind about the required work ?

File: src/rcc.rs [hunk 2]
```diff
@@ -80,13 +78,10 @@ impl APB1 {
...
 }
 
-
 impl APB1 {
     /// Set power interface clock (PWREN) bit in RCC_APB1ENR
     pub fn set_pwren(&mut self) {
-        self.enr().modify(|_r, w| {
-            w.pwren().set_bit()
-        })
+        self.enr().modify(|_r, w| w.pwren().set_bit())
```
Comment (geomatsi) [inline src/rcc.rs]: It looks like your commit includes ```cargo fmt``` formatting changes. IMHO it makes sense to keep coding style changes in a separate commit just to simplify review.

File: src/timer.rs (hunk 1 above)
Comment (cs2dsb) [reply src/timer.rs]: Will do.

File: src/rcc.rs (hunk 2 above)
Comment (cs2dsb) [reply src/rcc.rs]: I'll back these out and update the PR shortly.

File: src/timer.rs (hunk 1 above)
Comment (cs2dsb) [reply src/timer.rs]: Done and pushed

File: src/rcc.rs (hunk 2 above)
Comment (cs2dsb) [reply src/rcc.rs]: Done and pushed

**Comment by cs2dsb** [comment]
I left the piddling little rustfmt changes from random files out but added the pwm_input formatting fix back in as a separate commit because it took me half an hour to tweak the indentation and I feel like it's worth keeping ;)

File: src/timer.rs [hunk 3]
```diff
@@ -269,16 +290,74 @@ macro_rules! hal {
...
+}
+
+#[cfg(feature = "medium")]
 hal! {
-    TIM2: (tim2, pclk1_tim, dbg_tim2_stop),
-    TIM3: (tim3, pclk1_tim, dbg_tim3_stop),
     TIM4: (tim4, pclk1_tim, dbg_tim4_stop),
 }
+
+#[cfg(feature = "HIGH")]
```
Comment (burrbull) [inline src/timer.rs]: Why uppercase here?

Comment (cs2dsb) [reply src/timer.rs]: Good spot, that's a mistake.

Comment (cs2dsb) [reply src/timer.rs]: Fixed and pushed.

File: src/timer.rs [hunk 4]
```diff
@@ -1,7 +1,28 @@
 //! # Timer
 
 use crate::hal::timer::{CountDown, Periodic};
-use crate::pac::{DBGMCU as DBG, TIM1, TIM2, TIM3, TIM4};
+use crate::pac::{DBGMCU as DBG, TIM2, TIM3};
+#[cfg(any(feature = "stm32f100", feature = "stm32f103", feature = "stm32f105",))]
+use crate::pac::TIM1;
+#[cfg(feature = "medium")]
+use crate::pac::TIM4;
+#[cfg(feature = "HIGH")]
```
Comment (burrbull) [inline src/timer.rs]: Same here.

Comment (cs2dsb) [reply src/timer.rs]: Fixed and pushed

**Comment by cs2dsb** [comment]
Rebased latest changes from master.

Is there anything else I need to address to get this merged? I'm going to have a go at implementing the various PWM input & output features for the advanced timers and want to make sure that PR is easy to merge too.

Thanks
Daniel

File: src/timer.rs [hunk 5]
```diff
@@ -269,16 +307,91 @@ macro_rules! hal {
...
+
+#[cfg(all(
+    feature = "stm32f103", 
+    feature = "high",
+))]
+hal! {
+    TIM8: (tim8, pclk2_tim, dbg_tim8_stop),
+}
+
+//TODO: restore these timers once stm32-rs has been updated
```
Comment (burrbull) [inline src/timer.rs]: `stm32-rs` 0.9 released and we now use it or you mean something other?

Comment (cs2dsb) [reply src/timer.rs]: That refers to the missing dbg_tim(9 to 14)_stop fields on dbgmcu. They haven't been implemented in stm32-rs yet.

File: src/lib.rs [hunk 6]
```diff
@@ -94,14 +94,21 @@
 
 #![no_std]
 
+// Some pac crates have fields specified in such a way that they are safe and other
+// have them unsafe (likely an SVD error that needs patching). Unsafe blocks for 
+// one device cause warnings for the safe devices. This disables that warning.
+#![allow(unused_unsafe)]
```
Comment (TheZoq2) [inline src/lib.rs]: I think this error should have been fixed by the latest PAC release, could you check that?

Comment (cs2dsb) [reply src/lib.rs]: There are still instances of mixed safe & unsafe registers. Examples:
stm32f100::tim1::aar::AAR_W::bits is safe
stm32f100::tim6::aar::AAR_W::bits is unsafe

I think the original comment still stands for now.

**Comment by burrbull** [comment]
I think it can be merged after conflicts to be resolved.

**Comment by cs2dsb** [comment]
Updated with latest master.
Note I had to modify the new start_master (#99) part of the macro because the general purpose timers 9 to 14 don't support master mode.

File: src/lib.rs (hunk 6 above)
Comment (TheZoq2) [reply src/lib.rs]: Depending on how many such warnings there are, I would prefer having the `#[allow]` thing on the lines/functions where it happens. That way we don't miss actually useful warnings
